packages/core/src/utils/password.ts

```diff
@@ -4,10 +4,113 @@ import { type PasswordPolicyChecker } from '@logto/core-kit';
 import { type User, UsersPasswordEncryptionMethod } from '@logto/schemas';
 import { condObject } from '@silverhand/essentials';
 import { argon2i } from 'hash-wasm';
+import { z } from 'zod';
 
 import RequestError from '#src/errors/RequestError/index.js';
 import assertThat from '#src/utils/assert-that.js';
 
+type LegacyPassword = {
+  // `openssl list -digest-algorithms` will display the available digest algorithms.
+  algorithm: string;
+  args: string[];
+  encryptedPassword: string;
+};
+
+function isLegacyHashAlgorithm(algorithm: string): boolean {
+  try {
+    crypto.createHash(algorithm);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+function isLegacyPassword(value: string): [string, string[], string] | undefined {
+  try {
+    const legacyPasswordSchema = z.tuple([
+      z.string(),
+      z.array(z.string()).refine((args) => args.includes('@'), {
+        message: "Password args must include '@' character",
+      }),
+      z.string(),
+    ]);
+
+    const parsed = legacyPasswordSchema.safeParse(JSON.parse(value));
+
+    if (parsed.success) {
+      const [algorithm, args, encryptedPassword] = parsed.data;
+      return [algorithm, args, encryptedPassword];
+    }
+
+    return undefined;
+  } catch {
+    return undefined;
```

Comment:
if eslint won't complain:

```suggestion
    return;
  } catch {
    return;
```